requestparsing/test_harness.go

```diff
@@ -0,0 +1,120 @@
+package requestparsing
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+)
+
+type fakeHandler struct {
+	// callback is an anonymous function that verifies properties of the HTTP
+	// request and response provided in the test cases.
+	callback func(*http.Request)
+}
+
+func (h *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.callback(r)
+	io.WriteString(w, "Hello world!")
+}
+
+type fakeListener struct {
+	closeOnce      sync.Once
+	connChannel    chan net.Conn
+	serverEndpoint io.Closer
+	clientEndpoint net.Conn
+}
+
+// newFakeListener creates an instance of fakeListener. This will pass requests
+// to the HTTP server as part of the testing harness.
+func newFakeListener() *fakeListener {
+	s2c, c2s := net.Pipe()
+	c := make(chan net.Conn, 1)
+	c <- s2c
+	return &fakeListener{
+		connChannel:    c,
+		serverEndpoint: s2c,
+		clientEndpoint: c2s,
+	}
+}
+
+// Accept passes a network connection to the HTTP server to
+// enable bidirectional communication with the client. It will return an error
+// if Accept is called after the listener was closed
+func (l *fakeListener) Accept() (net.Conn, error) {
+	ch, ok := <-l.connChannel
+	if !ok {
+		return nil, errors.New("Listener closed")
+	}
+	return ch, nil
+}
+
+// Close will close the two network connections and the listener
+func (l *fakeListener) Close() (err error) {
+	l.closeOnce.Do(func() {
```

Comment:
This is a bit tricky.

On one hand, you are trying to use the named return value to save yourself creating a new variable and capture whatever is done in `Do()`, but you are also potentially leaving one endpoint open, even though it could've been closed.

Please find a way to refactor this.

Reply:
Refactored in 76e7795